### ISSUE TITLE
benchmarks

### ISSUE THREAD

**Issue by nickdesaulniers** [zlib-ng/zlib-ng#40]
I was just looking into pulling in zlib for a project today, and saw this on HN.  I'm curious, do you have any numbers that show performance improvements over vanilla zlib?  It's one thing to claim you're faster, but I didn't see any numbers in the readme.  I'm well aware benchmarks can be misleading, but profile profile profile!

**Comment by Dead2** [comment]
We have been doing some pretty rough cleanups of the codebase lately, so nobody has been looking too seriously into performance lately, except trying not to introduce big performance regressions (such as the minigzip decompress speed regression).

One of the more challenging problems right now is the buffer size problem (see issue #4) that causes zlib-ng to be incompatible with zlib due to needing SSE alignment. Workarounds are planned but we have prioritized getting the code and build system cleaned up first.

We also have to make sure an optimization for x86_64 does not regress on x86, arm, power, etc. This is quite a challenge at times since often optimizing for one will need really ugly workarounds (ifdefs and such) in order to not regress on the other architectures. Then add different compilers for each architecture and it quickly becomes a huge tangled mess.

So, to sum up, zlib-ng is not ready for mainstream use but we very much would like more contributors.

**Comment by nemequ** [comment]
FWIW if someone wants to create a [zlib-ng plugin for Squash](https://github.com/quixdb/squash/issues/108) then it would be included in the [Squash Benchmark](https://quixdb.github.io/squash-benchmark/).  I'll probably do it eventually, but it might be a while.

**Comment by nemequ** [comment]
I ended up adding a zlib-ng plugin to Squash, and I just released the first version of the [library](https://quixdb.github.io/squash/) (version 0.7.0) and [benchmark](https://quixdb.github.io/squash-benchmark/) which includes it.  Performance is generally a bit better than zlib, but it's a pretty small improvement.

You can hide items in the scatter plots by clicking on their labels in the legend.  Or, here is [a shortcut to only zlib and zlib-ng](https://quixdb.github.io/squash-benchmark/?visible-plugins=zlib,zlib-ng).

**Comment by Dead2** [comment]
Looks very interesting, awesome visual comparison tool indeed.

However, I wonder whether you set similar defines as running ./configure would to optimize for your platform. If you do not, then most optimizations will not be activated at all, thus staying close to zlib stock code and performance. Especially the level 1 compression should be noticeably different from stock zlib.

The CMake files have not been updated to feature all the checks needed to enable extra features yet, patches exist to do this but they have not been submitted yet.

If you use X86_64 on a cpu supporting SSE4.2 and use a recent GCC version, then these compiler settings/defines should be good:

```
-O2 -m64 -DZLIB_COMPAT -DWITH_GZFILEOP -DHAVE_HIDDEN -DHAVE_BUILTIN_CTZL -DMEDIUM_STRATEGY -DX86_64 -DX86_NOCHECK_SSE2 -DUNALIGNED_OK -DUNROLL_LESS -DX86_CPUID -DX86_SSE2_FILL_WINDOW -DX86_SSE4_2_CRC_HASH -DX86_PCLMULQDQ_CRC -DX86_QUICK_STRATEGY
```

For any of you following this project, you might have noticed a decline in activity the past few months. Mainly this is due to me being swamped with a very high priority project at work whilst also desperately trying to finish up things that need to be done before winter at home, combine that with family and you have pretty much no time and/or energy left.
I do have plans though and I hope to get back to zlib-ng soon-ish, but it might be weeks/months still.

**Comment by nemequ** [comment]
> However, I wonder whether you set similar defines as running ./configure would to optimize for your platform. If you do not, then most optimizations will not be activated at all, thus staying close to zlib stock code and performance. Especially the level 1 compression should be noticeably different from stock zlib.

Ah, _very_ good to know.  No, I don't currently do anything particularly special in the configuration, I'll look into that.

Obviously I can't just set those flags blindly; I could just copy the tests that are in zlib-ng currently, but how would you feel about creating a file which does all this stuff in zlib-ng, then both the main CMakeLists.txt in zlib-ng and the one from the Squash plugin could include it?  That would help anyone who is embedding zlib-ng in their project and keep their build system in sync, at least if they are using CMake…

Looking through the list of defines, I do have concerns.
- `-m64` is only useful for cross-compiling.
- `-DX86_64` sounds suspiciously like you're reinventing [existing pre-defined macros](http://sourceforge.net/p/predef/wiki/Architectures/#amd64)
- SSE support should really be determined at runtime instead of compile time so a single build can support both configurations.  Most people don't build their own software, and most distros and other packagers have to support CPUs without those extensions.
- I would be very wary of enabling depending on unaligned access.  It's very fragile and difficult to test reliably; just because a CPU allows unaligned access for some instructions doesn't mean it allows them for all, and new versions of compilers can start emitting more optimized instructions which don't allow unaligned access… GCC 4.9 did this and it broke some compression codecs (including LZ4).
- `-DHAVE_BUILTIN_CTZL` could just [check the GCC version](http://sourceforge.net/p/predef/wiki/Compilers/#gcc-cc) (`__builtin_ctzl` has been around since [version 3.4](https://gcc.gnu.org/onlinedocs/gcc-3.4.6/gcc/Other-Builtins.html)).

I also have some suspicions about `-DHAVE_VISIBILITY`, but I would have to take a look at what you're actually doing with it.  A hint, though: I don't want zlib-ng's symbols to be visible in the plugin.

**Comment by Dead2** [comment]
-m64 is not needed, these settings were just copied from one of my build logs.

SSE is runtime detected, unless you define -DX86_NOCHECK_SSE2, as that will assume that SSE2 is available (it always is on X86_64). I think all the rest of the instruction-set reliant features are runtime-checked and avoided if missing. But the defines needs to be set to include the features in the code at all. Unaligned access define is just one specific optimization for architectures that cannot do unaligned access, so setting the define disables the workaround.

HAVE_VISIBILITY is to enable compiling on really old compilers basically.

What -DX86_64 does, I cannot remember at the moment, but I think it has to do with both multilib and support for lots of different compilers. It could probably be done differently, and you are welcome to contribute improvements.

So, as long as you are on a X86_64 platform, all of those defines should be safe to set, even those that enable features for instructions you do not have, since it includes the code but does not force them to be used.

**Comment by Dead2** [comment]
The current proper way of getting these flags set is running ./configure, since that makes the proper makefile and sets up arch support and such. If you can run that before compiling, then that would currently be the best way to auto-detect flags. More complete CMake support is coming at some point.
